src/gedi_subset/maapx.py

```diff
@@ -2,109 +2,188 @@
+        408,  # Request Timeout
+        429,  # Too Many Requests
+        500,  # Internal Server Error
+        502,  # Bad Gateway
+        503,  # Service Unavailable
+        504,  # Gateway Timeout
+    }
+
+    if isinstance(e, requests.exceptions.RequestException):
+        status_code = e.response.status_code if e.response is not None else 0
+        return status_code not in _RETRYABLE_STATUS_CODES
+
+    return True
+
+
+def _make_retrier(max_tries=10) -> Callable[[Callable[P, T]], Callable[P, T]]:
+    return backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        giveup=_fatal_request_error,
+        max_tries=max_tries,
+        jitter=lambda wait: random.uniform(0.8 * wait, 1.2 * wait),
+        factor=10,
+    )
+
+
+def download_granule(
+    maap: MAAP,
+    todir: str,
+    granule: Granule,
+    *,
+    max_tries=10,
+) -> str | Never:
     """Download a granule's data file.
 
     Automatically fetch S3 credentials appropriate for `granule`, based upon
     it's S3 URL, and automatically refreshes credentials before expiry.
 
-    Return `IOSuccess[str]` containing the absolute path of the downloaded file
-    upon success; otherwise return `IOFailure[Exception]` containing the reason
-    for failure.
+    Parameters
+    ----------
+    maap
+        The MAAP client to use for downloading the granule.
+    todir
+        The directory to which to download the granule.
+    granule
+        The metadata for the granule to download.
+
+    Returns
+    -------
+    dest
+        The absolute path of the downloaded file.
+
+    Raises
+    ------
+    ValueError
+        If the granule is not configured with a download URL.
     """
     granule_ur = granule["Granule"]["GranuleUR"]
-    download_url = granule.getDownloadUrl()
-    logger.debug(f"Downloading granule {granule_ur} to directory {todir}")
-
-    if download_url and download_url.startswith("s3"):
-        result: Union[Maybe, IOResultE] = flow(
-            _s3_credentials_endpoint(granule),
-            bind(partial(_earthdata_s3_credentials, maap)),
-            map_(_setup_default_boto3_session),
-        )
-
-        # If result is an IOResult, that means the granule has an S3 credentials
-        # endpoint, and an attempt was made to obtain S3 credentials.  If the
-        # attempt was unsuccessful, return the failure rather than proceeding to
-        # download the granule.
-        if isinstance(result, IOResult) and not is_successful(result):
-            return result
-
-    return impure_safe(granule.getData)(todir)
+    download_url = granule.getDownloadUrl() or ""
+    logger.debug(f"Downloading granule {granule_ur} from {download_url} to {todir}")
+
+    if download_url.startswith("s3"):
+        if endpoint := _s3_credentials_endpoint(granule):
+            _setup_default_boto3_session(_s3_credentials(maap, endpoint))
+
+    retry_get_data = _make_retrier(max_tries)(granule.getData)
```

Comment:
`granule.getData` is the method that downloads the granule's data file. Here, we're simply decorating the `getData` method with the retrier, which will catch errors, examine the type of error to see if it is a retriable type, and either retry the `getData` call or propagate the error.

Reply:
For further clarity, this expression:

```python
_make_retrier(max_tries)(granule.getData)
```

is roughly equivalent to this:

```python
@_make_retrier()
def getData(self, dir: str):
    ...
```

However, we can neither use a decorator like that since the source code for `getData` is not within this repo, nor would that standard approach to decorating a function allow us to dynamically pass a `max_tries` value to alter configuration, which is an important thing to allow, particularly for unit testing, if we don't want to deal with some ugly mocking.